src/index.js

```diff
@@ -165,5 +180,71 @@ export default class ScrollPlugin extends Plugin {
                 swup.scrollTo(0, this.shouldAnimate('betweenPages'));
             }
         }
-    };
+    }
+
+    onWillReplaceContent = () => {
+        this.storeScrollPositions(this.previousUrl);
+        this.previousUrl = window.location.href;
+    }
+
+    onClickLink = (e) => {
+        this.deleteStoredScrollPositions(e.delegateTarget.href);
+    }
+
+    /**
+     * Stores the scroll positions for the current URL
+     * @param {string} url 
+     */
+    storeScrollPositions(url) {
+      // create an object to store the scroll positions
+      const storeEntry = {
+        window: {top: window.scrollY, left: window.scrollX},
+        elements: []
+      }
+      // fill up the object with scroll positions for each matching element
+      const $elements = document.querySelectorAll('[data-swup-restore-scroll]');
+      $elements.forEach(el => storeEntry.elements.push({
+        top: el.scrollTop, 
+        left: el.scrollLeft
+      }));
+      // put the object into the store
+      this.scrollPositionsStore[url] = storeEntry;
+    }
+
+    /**
+     * Deletes stored scroll positions for a given URL
+     * @param {string} url 
+     */
+    deleteStoredScrollPositions(url) {
+      delete this.scrollPositionsStore[url];
+      this.scrollPositionsStore[url] = null;
+    }
+
+    /**
+     * Restore the scroll positions for all matching elements
+     * @returns void
+     */
+    restoreScrollPositions() {
+      const swup = this.swup;
+      // get the stored scroll positions from the cache
+      const scrollPositions = this.scrollPositionsStore[window.location.href];
+      if (scrollPositions == null) {
+        return;
+      }
+      if (scrollPositions.elements == null) {
+        return;
+      }
+      // cycle through all elements on the current page and restore their scroll positions, if appropriate
+      const $elements = document.querySelectorAll('[data-swup-restore-scroll]');
```

Comment:
I'd use the `queryAll` helper here, it converts the NodeList to an array which I think Safari still has problems with.

Reply:
Oops really? Do you have a source describing the Safari issue? If so, I have been building a lot of broken websites in the last few years 😅

Reply:
Mmh. Officially [it's supported from Safari 10](https://developer.mozilla.org/en-US/docs/Web/API/NodeList/forEach#browser_compatibility), so that should never have been an issue. Maybe I'm misremembering and it was Edge? It definitely was an issue as recently as 1/2 years ago, even with Babel.

Reply:
As far as I know, [we should be fine](https://caniuse.com/queryselector)

Reply:
Do you know how I could actually access swups `queryAll` helper function from the plugin? I don't really care, would just as well use that instead of `querySelectorAll`, if I would have access to it ;)